pallets/parachain-staking/src/set.rs

```diff
@@ -124,6 +126,28 @@ impl<T: Ord> OrderedSet<T> {
 		self.0.binary_search(value)
 	}
 
+	// iterative search from left to right which is assumed to be sorted from
+	// greatest to least
+	pub fn linear_search(&self, value: &T) -> Result<usize, usize> {
+		let size = self.0.len();
+		let mut loc: usize = size;
+		let mut i: usize = 0;
+		// keep running until we find a smaller item
+		for v in self.0.iter() {
+			// prevent to have same items
+			if v.cmp(value) == Ordering::Equal {
+				return Ok(i);
+			// store index if v is bigger (since we assume order from max to
+			// min) but keep checking for Ordering::Equal
+			} else if v.cmp(value) == Ordering::Less && loc == size {
+				// insert after current element
+				loc = i;
+			}
+			i = i.saturating_add(1);
+		}
```

Comment:
No need to keep track of i manually! You could use [enumerate](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate).

Something like that should work too?

```suggestion
		self.0.iter().enumerate().find_map(|(i, elem)| {
			// prevent to have same items
			match v.cmp(value) {
			Ordering::Equal => Some(Ok(i))
			// store index if v is bigger (since we assume order from max to
			// min) but keep checking for Ordering::Equal
			Ordering::Less => {
				// insert after current element
				Some(Err(i))
			},
			_ => None,
		})
```

I don't really get the `else if` part. Wouldn't we want to return there?

Reply:
> I don't really get the `else if` part. Wouldn't we want to return there?

I had that in the beginning but this does not work unfortunately because then we do not check whether the account id is already in the set. Thus, we end up with account ids existing more than once in the set.

Check the following example from the OrderedSet test `linear_search` with your suggestion, e.g., returning in the `else if` part:

```
OrderedSet::from_sorted(vec![
 { owner: 1, amount: 100 },
 { owner: 3, amount: 90 },
 { owner: 5, amount: 80 },
 { owner: 2, amount: 75 },
 { owner: 7, amount: 70 },
 { owner: 9, amount: 60 }
]).linear_search({ owner: 2, amount: 90}) = Some(Err(2))
```

However, it should return `Some(Ok(3))`, change the value and then resort.

Reply:
Should be fixed in 211fabd